src/Controller/CaseEventController.php

```diff
@@ -0,0 +1,172 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use Symfony\Component\Translation\TranslatableMessage;
+
+#[Route('/case/{id}/case-events')]
+class CaseEventController extends AbstractController
+{
+    #[Route('/', name: 'case_event_index', methods: ['GET'])]
+    public function index(CaseEntity $case, CaseEventRepository $caseEventRepository, FilterBuilderUpdaterInterface $filterBuilderUpdater, PaginatorInterface $paginator, Request $request): Response
+    {
+        $filterOptions = [
+            'case' => $case,
+            'method' => 'get',
+            'action' => $this->generateUrl('case_event_index', ['id' => $case->getId()]),
+        ];
+
+        $filterForm = $this->createForm(CaseEventFilterType::class, null, $filterOptions);
+        if ($request->query->has($filterForm->getName())) {
+            $filterForm->submit($request->query->get($filterForm->getName()));
+        }
+
+        $filterBuilder = $caseEventRepository->createAvailableCaseEventsForCaseQueryBuilder('ce', $case);
+        $filterBuilderUpdater->addFilterConditions($filterForm, $filterBuilder);
+        $query = $filterBuilder->getQuery();
+
+        $pagination = $paginator->paginate(
+            $query,
+            1,
+            1000, // Hopefully this is never reached.
+            [
+                'defaultSortFieldName' => 'ce.receivedAt',
+                'defaultSortDirection' => Criteria::DESC,
+            ]
+        );
+
+        return $this->render('case/event/index.html.twig', [
+            'filter_form' => $filterForm->createView(),
+            'case' => $case,
+            'pagination' => $pagination,
+        ]);
+    }
+
+    #[Route('/{caseEvent}/show', name: 'case_event_show', methods: ['GET'])]
+    public function show(CaseEntity $case, CaseEvent $caseEvent): Response
+    {
+        return $this->render('case/event/show.html.twig', [
+            'case' => $case,
+            'case_event' => $caseEvent,
+        ]);
+    }
+
+    #[Route('/create', name: 'case_event_create', methods: ['GET', 'POST'])]
+    public function create(CaseEntity $case, EntityManagerInterface $manager, PartyHelper $partyHelper, Request $request): Response
+    {
+        if (!($this->isGranted('ROLE_CASEWORKER') || $this->isGranted('ROLE_ADMINISTRATION'))) {
+            throw new AccessDeniedException();
+        }
+
+        $parties = $partyHelper->getRelevantPartiesByCase($case);
+
+        // Convert into choices for CaseEventNewType.
+        $choices = [];
+        foreach ($parties['parties'] as $data) {
+            /** @var Party $party */
+            $party = $data['party'];
+            $type = $data['type'];
+            $choices[$party->getName().', '.$type] = $party;
+        }
+
+        foreach ($parties['counterparties'] as $data) {
```

Comment:
Are we sure that `$parties['parties']` and `$parties['counterparties']` are always set?

Reply:
They might very well be empty, but yes they will always be there.